api/paidAction/index.js

```diff
@@ -0,0 +1,310 @@
+        id: invoiceId,
+        actionState: 'FAILED'
+      },
+      data: {
+        actionState: 'RETRYING'
+      }
+    })
+
+    context.cost = BigInt(msatsRequested)
+    context.actionId = actionId
+
+    // create a new invoice
+    const invoice = await createDbInvoice(actionType, args, context)
+
+    return {
+      result: await action.retry({ invoiceId, newInvoiceId: invoice.id }, context),
+      invoice,
+      paymentMethod: 'OPTIMISTIC'
+    }
+  }, { isolationLevel: Prisma.TransactionIsolationLevel.ReadCommitted })
+}
+
+const OPTIMISTIC_INVOICE_EXPIRE = { minutes: 10 }
+const PESSIMISTIC_INVOICE_EXPIRE = { minutes: 10 }
+
+async function createDbInvoice (actionType, args, context) {
+  const { user, models, tx, lnd, cost, optimistic, actionId } = context
+  const action = paidActions[actionType]
+  const db = tx ?? models
+  const [createLNDInvoice, expirePivot, actionState] = optimistic
+    ? [createInvoice, OPTIMISTIC_INVOICE_EXPIRE, 'PENDING']
+    : [createHodlInvoice, PESSIMISTIC_INVOICE_EXPIRE, 'PENDING_HELD']
+
+  if (cost < 1000n) {
+    // sanity check
+    throw new Error('The cost of the action must be at least 1 sat')
+  }
+
+  const expiresAt = datePivot(new Date(), expirePivot)
+  const lndInv = await createLNDInvoice({
+    description: user?.hideInvoiceDesc ? undefined : await action.describe(args, context),
+    lnd,
+    mtokens: String(cost),
+    expires_at: expiresAt
+  })
+
+  const invoice = await db.invoice.create({
+    data: {
+      hash: lndInv.id,
+      msatsRequested: cost,
+      preimage: optimistic ? undefined : lndInv.secret,
+      bolt11: lndInv.request,
+      userId: user?.id || USER_ID.anon,
+      actionType,
+      actionState,
+      expiresAt,
+      actionId
+    }
+  })
+
+  // insert a job to check the invoice after it's set to expire
+  await db.$executeRaw`
+      INSERT INTO pgboss.job (name, data, retrylimit, retrybackoff, startafter, expirein, priority)
+      VALUES ('checkInvoice',
+        jsonb_build_object('hash', ${lndInv.id}::TEXT), 21, true,
+          ${expiresAt}::TIMESTAMP WITH TIME ZONE,
+          ${expiresAt}::TIMESTAMP WITH TIME ZONE - now() + interval '10m', 100)`
+
+  // the HMAC is only returned during invoice creation
+  // this makes sure that only the person who created this invoice
+  // has access to the HMAC
+  invoice.hmac = createHmac(invoice.hash)
+
+  return invoice
+}
+
+export async function verifyPayment ({ hash, hmac, models, cost }) {
+  if (!hash) {
+    throw new Error('hash required')
+  }
+
+  if (!hmac) {
+    throw new Error('hmac required')
+  }
+
+  const hmac2 = createHmac(hash)
+  if (!timingSafeEqual(Buffer.from(hmac), Buffer.from(hmac2))) {
+    throw new Error('hmac invalid')
+  }
+
+  const invoice = await models.invoice.findUnique({
+    where: {
+      hash,
+      actionState: 'HELD'
+    }
+  })
+
+  if (invoice.msatsReceived < cost) {
+    throw new Error('invoice amount too low')
+  }
+
+  return invoice
+}
```

Comment:
This function looks suspiciously simple compared to all the checks in the previous version :monocle_face: 

But I guess making sure it's in `HELD` makes all the other checks indeed obsolete.